File: I1_MVPAndRetrofit/app/src/main/java/com/example/tsvetelinastoyanova/weatherreportrevisited/MainActivity.java
package com.example.tsvetelinastoyanova.weatherreportrevisited;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import com.example.tsvetelinastoyanova.weatherreportrevisited.citieslist.CitiesListFragment;
import com.example.tsvetelinastoyanova.weatherreportrevisited.citieslist.CitiesListPresenter;
import com.example.tsvetelinastoyanova.weatherreportrevisited.model.WeatherObject;
import com.example.tsvetelinastoyanova.weatherreportrevisited.util.Utils;
import com.example.tsvetelinastoyanova.weatherreportrevisited.weatherdetails.WeatherDetailsFragment;
import com.example.tsvetelinastoyanova.weatherreportrevisited.weatherdetails.WeatherDetailsPresenter;

public class MainActivity extends AppCompatActivity implements CitiesListFragment.OnClickCityDelegate {

    private WeatherDetailsFragment weatherDetailsFragment;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        if (findViewById(R.id.list_fragment_container) != null) {
            createCitiesListFragment();
        }
        if (findViewById(R.id.details_fragment_container) != null) {
            createWeatherListFragment();
        }
    }

    @Override
    public void onClickCity(WeatherObject weatherObject) {
        if (Utils.isTablet(this)) {
            weatherDetailsFragment.changeView(weatherObject);
        } else {
            Intent i = new Intent(MainActivity.this, WeatherDetailsActivity.class);
            i.putExtra(Constants.WEATHER_OBJECTS, weatherObject);
            startActivity(i);
        }
    }

    private void createCitiesListFragment() {
        CitiesListFragment citiesListFragment =
                (CitiesListFragment) getSupportFragmentManager().findFragmentById(R.id.list_fragment_container);
        if (citiesListFragment == null) {
            // Create the fragment
            citiesListFragment = CitiesListFragment.newInstance();
            CitiesListPresenter citiesListPresenter = new CitiesListPresenter(citiesListFragment, this.getApplicationContext(), Utils.provideCityRepository(getApplicationContext()));
            citiesListFragment.setPresenter(citiesListPresenter);
            Utils.addFragmentToActivity(
                    getSupportFragmentManager(), citiesListFragment, R.id.list_fragment_container);
        }
    }

    private void createWeatherListFragment() {
        weatherDetailsFragment =
                (WeatherDetailsFragment) getSupportFragmentManager().findFragmentById(R.id.details_fragment_container);
        if (weatherDetailsFragment == null) {
            // Create the fragment
            weatherDetailsFragment = WeatherDetailsFragment.newInstance();
            WeatherDetailsPresenter weatherDetailsPresenter = new WeatherDetailsPresenter();
            weatherDetailsFragment.setPresenter(weatherDetailsPresenter);
            Utils.addFragmentToActivity(
                    getSupportFragmentManager(), weatherDetailsFragment, R.id.details_fragment_container);
        }
    }

}

File: I1_MVPAndRetrofit_Kotlin/app/src/main/java/com/example/tsvetelinastoyanova/weatherapp/MainActivity.kt
package com.example.tsvetelinastoyanova.weatherapp

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import com.example.tsvetelinastoyanova.weatherapp.citiesList.CitiesListFragment
import com.example.tsvetelinastoyanova.weatherapp.citiesList.CitiesListPresenter
import com.example.tsvetelinastoyanova.weatherapp.data.source.CitiesRepository
import com.example.tsvetelinastoyanova.weatherapp.model.currentweather.CurrentWeatherObject
import com.example.tsvetelinastoyanova.weatherapp.util.Utils
import com.example.tsvetelinastoyanova.weatherapp.weatherdetails.WeatherDetailsContainerFragment
import com.example.tsvetelinastoyanova.weatherapp.weatherdetails.WeatherDetailsContainerPresenter
import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity(), CitiesListFragment.OnClickCityDelegate {

    private var weatherDetailsContainerFragment: WeatherDetailsContainerFragment? = null
    private var weatherObject: CurrentWeatherObject? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        weatherObject = savedInstanceState?.getParcelable(Constants.WEATHER_OBJECTS)

        if (list_fragment_container != null) {
            createCitiesListFragment()
        }
        if (details_fragment_container != null) {
            createWeatherDetailsFragment()
        }
    }

    override fun onClickCity(currentWeatherObject: CurrentWeatherObject) {
        if (Utils.isTablet(this)) {
            weatherObject = currentWeatherObject
            weatherDetailsContainerFragment?.clickOnCity(currentWeatherObject, true)
        } else {
            val i = Intent(this@MainActivity, WeatherDetailsActivity::class.java)
            i.putExtra(Constants.WEATHER_OBJECTS, currentWeatherObject)
            startActivity(i)
        }
    }

    private fun createCitiesListFragment() {
        var citiesListFragment: CitiesListFragment? = supportFragmentManager.findFragmentById(R.id.list_fragment_container) as? CitiesListFragment
        if (citiesListFragment == null) {
            citiesListFragment = CitiesListFragment.newInstance()
            Utils.addFragmentToActivity(
                supportFragmentManager, citiesListFragment, R.id.list_fragment_container)
        }
        val citiesRepository: CitiesRepository? = Utils.provideCityRepository(applicationContext)
        citiesRepository?.let {
            val citiesListPresenter = CitiesListPresenter(citiesListFragment, it)
            citiesListFragment.setPresenter(citiesListPresenter)
        }
    }

    private fun createWeatherDetailsFragment() {
        var weatherDetailsContainerFragment: WeatherDetailsContainerFragment? = supportFragmentManager.findFragmentById(R.id.details_fragment_container) as? WeatherDetailsContainerFragment
        if (weatherDetailsContainerFragment == null) {
            weatherDetailsContainerFragment = WeatherDetailsContainerFragment.newInstance()
            Utils.addFragmentToActivity(
                supportFragmentManager, weatherDetailsContainerFragment, R.id.details_fragment_container)
        }
        val weatherDetailsContainerPresenter = WeatherDetailsContainerPresenter(weatherDetailsContainerFragment, Utils.provideCityRepository(this))
        weatherDetailsContainerFragment.setPresenter(weatherDetailsContainerPresenter)
        this.weatherDetailsContainerFragment = weatherDetailsContainerFragment

        weatherObject?.let {
            onClickCity(it)
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        outState?.run { putParcelable(Constants.WEATHER_OBJECTS, weatherObject) }
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
        super.onRestoreInstanceState(savedInstanceState)
        weatherObject = savedInstanceState?.getParcelable(Constants.WEATHER_OBJECTS)
    }
}
